modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/TransportSearchTemplateAction.java

```diff
@@ -71,18 +71,29 @@ public TransportSearchTemplateAction(
     @Override
     protected void doExecute(Task task, SearchTemplateRequest request, ActionListener<SearchTemplateResponse> listener) {
         final SearchTemplateResponse response = new SearchTemplateResponse();
+        boolean success = false;
         try {
             SearchRequest searchRequest = convert(request, response, scriptService, xContentRegistry, searchUsageHolder);
             if (searchRequest != null) {
-                client.search(searchRequest, listener.delegateFailureAndWrap((l, searchResponse) -> {
+                client.search(searchRequest, listener.delegateResponse((l, e) -> {
+                    response.decRef();
```

Comment:
yuck admittedly, but no matter how I do this it looks horrible IMO, the problem is that we reference (+ mutate!) the response in both the method context and the listener and have to make sure we properly release no matter what exception we run into ... maybe we can improve the way this works in the first place but for now this was the version I least disliked ...

Reply:
I can't see a better way 🤷